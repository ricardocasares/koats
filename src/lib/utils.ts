export const sleep = (n: number) =>
  new Promise(resolve => setTimeout(resolve, n));
