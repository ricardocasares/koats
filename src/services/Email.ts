import { Email } from "@/models";

export class EmailService {
  send(name: string): Promise<Email> {
    return Promise.resolve({ id: 1, to: name, sent: true });
  }
}
