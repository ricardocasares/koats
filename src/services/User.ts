import { User } from "@/models";

export class UserService {
  id(id: number): Promise<User> {
    return Promise.resolve({ id, email: "meas@app.io", locked: true });
  }

  email(email: string): Promise<User> {
    return Promise.resolve({ id: 1, email: "pete", locked: true });
  }

  unlock(id: number): Promise<User> {
    return Promise.resolve({ id, email: "measd@app.io", locked: false });
  }
}
