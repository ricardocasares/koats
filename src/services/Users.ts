export type IUser = {
  id: number;
  email: string;
  locked: boolean;
};

export class Users {
  id(id: number): Promise<IUser> {
    return Promise.resolve({ id, email: "me@app.io", locked: true });
  }

  email(email: string): Promise<IUser> {
    return Promise.resolve({ id: 1, email, locked: true });
  }

  unlock(id: number): Promise<IUser> {
    return Promise.resolve({ id, email: "me@app.io", locked: false });
  }
}
