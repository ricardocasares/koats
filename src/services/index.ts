export * from "./Users";
export * from "./Email";
export * from "./Logger";
