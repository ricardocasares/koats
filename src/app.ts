import kh from "koa-helmet";
import pino from "koa-pino-logger";

import { app } from "@/koa";
import { logger } from "@/lib/logger";
import { router } from "@/router";
import { errors } from "@/middleware/errors";
import { tracer } from "@/middleware/tracer";
import { Dependencies } from "@/models";

export const createApp = (deps: Dependencies) => {
  // set dependencies
  app.context.dc = deps;
  // add middleware
  return app
    .use(kh())
    .use(
      tracer({
        config: {
          serviceName: "koats",
          sampler: {
            type: "const",
            param: 1.0
          }
        },
        options: { logger, tags: { version: "1.0.0" } }
      })
    )
    .use(errors)
    .use(pino({ logger }))
    .use(router.routes());
};
