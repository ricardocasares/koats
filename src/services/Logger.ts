import pino, { LoggerOptions } from "pino";
import { isTest } from "@/lib/env";

export const Logger = pino({
  name: "csa",
  enabled: !isTest(),
  level: "info"
});

export default Logger;
