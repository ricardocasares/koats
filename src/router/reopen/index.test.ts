import request from "supertest";
import { createApp } from "@/app";
import { UserService as User, EmailService as Email } from "@/services";
import { snapshot } from "@/test/utils";

// dependencies
const users = new User();
const emails = new Email();

// mocked data
const user = { id: 1, email: "me@app.io", locked: true };
const email = { id: 1, to: "me@app.io", sent: true };

// init application
const app = createApp({ users, emails });

describe("/reopen", () => {
  it("should reopen a given account", async () => {
    jest.spyOn<User, "id">(users, "id").mockResolvedValue(user);
    jest.spyOn<User, "email">(users, "email").mockResolvedValue(user);
    jest.spyOn<Email, "send">(emails, "send").mockResolvedValue(email);

    await request(app.callback())
      .get("/reopen/1")
      .expect(snapshot);
  });

  it("should find an account by email as failover", async () => {
    jest.spyOn<User, "id">(users, "id").mockRejectedValue(new Error());
    jest.spyOn<User, "email">(users, "email").mockResolvedValue(user);
    jest.spyOn<Email, "send">(emails, "send").mockResolvedValue(email);

    await request(app.callback())
      .get("/reopen/1")
      .expect(snapshot);
  });

  it("should fail when no account is found", async () => {
    jest.spyOn<User, "id">(users, "id").mockRejectedValue(new Error());
    jest.spyOn<User, "email">(users, "email").mockRejectedValue(new Error());

    await request(app.callback())
      .get("/reopen/1")
      .expect(snapshot);
  });
});
