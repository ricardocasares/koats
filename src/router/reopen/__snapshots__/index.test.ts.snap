// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/reopen should fail when no account is found 1`] = `
Object {
  "body": Object {
    "message": "Account not found",
    "state": Object {},
    "status": 404,
  },
  "headers": Object {
    "connection": "close",
    "content-length": "55",
    "content-type": "application/json; charset=utf-8",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-xss-protection": "1; mode=block",
  },
  "status": 404,
}
`;

exports[`/reopen should find an account by email as failover 1`] = `
Object {
  "body": Object {
    "message": "response sent",
    "state": Object {
      "account": Object {
        "email": "me@app.io",
        "id": 1,
        "locked": false,
      },
      "email": Object {
        "id": 1,
        "sent": true,
        "to": "me@app.io",
      },
    },
  },
  "headers": Object {
    "connection": "close",
    "content-length": "137",
    "content-type": "application/json; charset=utf-8",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-xss-protection": "1; mode=block",
  },
  "status": 200,
}
`;

exports[`/reopen should reopen a given account 1`] = `
Object {
  "body": Object {
    "message": "response sent",
    "state": Object {
      "account": Object {
        "email": "me@app.io",
        "id": 1,
        "locked": false,
      },
      "email": Object {
        "id": 1,
        "sent": true,
        "to": "me@app.io",
      },
    },
  },
  "headers": Object {
    "connection": "close",
    "content-length": "137",
    "content-type": "application/json; charset=utf-8",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-xss-protection": "1; mode=block",
  },
  "status": 200,
}
`;
